File: LNCollectionView/LNCollectionView/LNCollectionView.m
//
//  LNCollectionView.m
//  LNCollectionView
//
//  Created by Levison on 13.11.24.
//

#import "LNCollectionView.h"
#import "LNCollectionViewCell.h"
#import "LNCollectionViewReusePool.h"

@interface LNCollectionViewLayout (UICollectionViewNeed)

@property (nullable, nonatomic) LNCollectionView *collectionView;

@end

@interface LNCollectionView ()

@property (nonatomic, strong) LNCollectionViewLayout *collectionViewLayout;

@property (nonatomic, copy) NSArray<LNCollectionViewLayoutAttributes *> *currentAttributesArr;
@property (nonatomic, strong) NSMutableDictionary<NSIndexPath *, LNCollectionViewCell *> *currentCells;

@property (nonatomic, assign) BOOL hasInitialized;

@property (nonatomic, strong) LNCollectionViewReusePool *pool;

@end

@implementation LNCollectionView

- (instancetype)initWithFrame:(CGRect)frame collectionViewLayout:(LNCollectionViewLayout *)layout
{
    self = [super initWithFrame:frame];
    if (self) {
        self.collectionViewLayout = layout;
        self.collectionViewLayout.collectionView = self;
    }
    return self;
}

- (void)setContentOffset:(CGPoint)contentOffset
{
    [super setContentOffset:contentOffset];
    [self checkVisible];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    //其他时机是reloadData
    if (!self.hasInitialized) {
        [self.collectionViewLayout prepareLayout];
        [self checkVisible];
        self.hasInitialized = YES;
    }
}

- (void)reloadData
{
    for (LNCollectionViewLayoutAttributes *attributes in self.currentAttributesArr) {
        LNCollectionViewCell *cell = [self.currentCells objectForKey:attributes.indexPath];
        [cell removeFromSuperview];
        [self.currentCells removeObjectForKey:attributes.indexPath];
        [self.pool addReusableCell:cell];
    }
    [self.currentCells removeAllObjects];
    self.currentAttributesArr = @[];
    [self.collectionViewLayout invalidateLayout];
    [self.collectionViewLayout prepareLayout];
    [self checkVisible];
}

- (void)checkVisible
{
    NSArray<LNCollectionViewLayoutAttributes *> *currentAttributesArr = [self.collectionViewLayout layoutAttributesForElementsInRect:self.bounds];
    NSSet<LNCollectionViewLayoutAttributes *> *visibleAttributesSet = [NSSet setWithArray:self.currentAttributesArr];
    NSSet<LNCollectionViewLayoutAttributes *> *newAttributesSet = [NSSet setWithArray:currentAttributesArr];
    NSMutableSet<LNCollectionViewLayoutAttributes *> *newlyVisibleAttributesMSet= [newAttributesSet mutableCopy];
    [newlyVisibleAttributesMSet minusSet:visibleAttributesSet];
    NSMutableSet<LNCollectionViewLayoutAttributes *> *disappearingAttributesMSet = [visibleAttributesSet mutableCopy];
    [disappearingAttributesMSet minusSet:newAttributesSet];
    for (LNCollectionViewLayoutAttributes *attributes in newlyVisibleAttributesMSet) {
        //willDisplay
        LNCollectionViewCell *cell = [self.dataSource ln_collectionView:self cellForItemAtIndexPath:attributes.indexPath];
        [self.currentCells setObject:cell forKey:attributes.indexPath];
        [self addSubview:cell];
        cell.frame = [self.collectionViewLayout layoutAttributesForItemAtIndexPath:attributes.indexPath].frame;
    }

    for (LNCollectionViewLayoutAttributes *attributes in disappearingAttributesMSet) {
        //willDisappear
        LNCollectionViewCell *cell = [self.currentCells objectForKey:attributes.indexPath];
        [cell removeFromSuperview];
        [self.currentCells removeObjectForKey:attributes.indexPath];
        [self.pool addReusableCell:cell];
    }
    self.currentAttributesArr = newAttributesSet.allObjects;
    self.contentSize = self.collectionViewLayout.collectionViewContentSize;
}

- (void)registerClass:(Class)cellClass forCellWithReuseIdentifier:(NSString *)identifier
{
    [self.pool registerClass:cellClass forCellWithReuseIdentifier:identifier];
}

- (__kindof LNCollectionViewCell *)dequeueReusableCellWithReuseIdentifier:(NSString *)identifier forIndexPath:(NSIndexPath *)indexPath
{
    return [self.pool dequeueReusableCellWithIdentifier:identifier];
}

- (NSMutableDictionary<NSIndexPath *, LNCollectionViewCell *> *)currentCells
{
    if (!_currentCells) {
        _currentCells = [[NSMutableDictionary alloc] init];
    }
    return _currentCells;
}

- (LNCollectionViewReusePool *)pool
{
    if (!_pool) {
        _pool = [[LNCollectionViewReusePool alloc] init];
    }
    return _pool;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/collectionview/LNCollectionView.kt
package com.ln.collectionview.collectionview

import android.content.Context
import android.graphics.Rect
import android.util.Log
import android.view.ViewGroup
import com.ln.collectionview.collectionview.attributes.LNCollectionViewLayoutAttributes
import com.ln.collectionview.collectionview.cell.LNCollectionViewCell
import com.ln.collectionview.collectionview.layout.LNCollectionViewLayout
import com.ln.collectionview.collectionview.layout.LNIndexPath
import com.ln.collectionview.collectionview.reuse.LNCollectionViewReusePool
import com.ln.collectionview.scrollview.LNScrollView
import com.ln.collectionview.scrollview.LNScrollViewDelegate
import java.lang.ref.WeakReference

interface LNCollectionViewDataSource {
    fun collectionViewNumberOfItemsInSection(collectionView: LNCollectionView, section: UInt): UInt
    fun collectionViewCellForItemAtIndexPath(collectionView: LNCollectionView, indexPath: LNIndexPath): LNCollectionViewCell
    fun numberOfSectionsInCollectionView(collectionView: LNCollectionView): UInt { return 1u }
}

interface LNCollectionViewDelegate : LNScrollViewDelegate {
    fun collectionViewDidSelectItemAtIndexPath(collectionView: LNCollectionView, indexPath : LNIndexPath) {}
    fun collectionViewDidDeselectItemAtIndexPath(collectionView: LNCollectionView, indexPath: LNIndexPath) {}
    fun collectionViewWillDisplayCellForItemAtIndexPath(collectionView: LNCollectionView, cell: LNCollectionViewCell, indexPath: LNIndexPath) {}
    fun collectionViewDidEndDisplayingCellForItemAtIndexPath(collectionView: LNCollectionView, cell: LNCollectionViewCell, indexPath: LNIndexPath) {}
}

class LNCollectionView(
    context: Context,
    var collectionViewLayout: LNCollectionViewLayout
) : LNScrollView(context) {

    var delegate: WeakReference<LNCollectionViewDelegate>? = null
    var dataSource: WeakReference<LNCollectionViewDataSource>? = null

    var currentAttributesArr: List<LNCollectionViewLayoutAttributes> = emptyList()
    var currentCells: MutableMap<String, LNCollectionViewCell> = mutableMapOf()
    var hasInitialized = false
    var pool: LNCollectionViewReusePool = LNCollectionViewReusePool(this)

    init {
        collectionViewLayout.collectionViewWeakRef = WeakReference<LNCollectionView>(this)
    }

    override fun setScrollX(value: Int) {
        super.setScrollX(value)
        checkVisible()
    }

    override fun setScrollY(value: Int) {
        super.setScrollY(value)
        checkVisible()
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        if (!hasInitialized) {
            collectionViewLayout.prepareLayout()
            checkVisible()
            hasInitialized = true
        }
    }

    fun reloadData() {
        for (attributes in currentAttributesArr) {
            val cell = currentCells[attributes.indexPath.key()]
            cell?.let {
                removeView(it)
                currentCells.remove(attributes.indexPath.key())
                pool.addReusableCell(it)
            }
        }
        currentCells.clear()
        currentAttributesArr = emptyList()
        collectionViewLayout.invalidateLayout()
        collectionViewLayout.prepareLayout()
        checkVisible()
    }

    private fun checkVisible() {
        val currentBounds = Rect(scrollX, scrollY, scrollX + width, scrollY + height)
        var newAttributesArr = collectionViewLayout.layoutAttributesForElementsInRect(currentBounds)
        val visibleAttributesSet = currentAttributesArr.toSet()
        val newAttributesSet = newAttributesArr.toSet()
        val newlyVisibleAttributesMSet = newAttributesSet.toMutableSet().apply { minusAssign(visibleAttributesSet) }
        val disappearingAttributesMSet = visibleAttributesSet.toMutableSet().apply { minusAssign(newAttributesSet) }


        if (dataSource?.get() == null) {
            return
        }
        val validDataSource = dataSource?.get()!!
        for (attributes in newlyVisibleAttributesMSet) {
            val cell = validDataSource.collectionViewCellForItemAtIndexPath(this, attributes.indexPath).let {
                currentCells[attributes.indexPath.key()] = it
                this@LNCollectionView.addView(it)
                it.layout(attributes.frame.left, attributes.frame.top, attributes.frame.right, attributes.frame.bottom)
                it
            }
        }

        for (attributes in disappearingAttributesMSet) {
            val cell = currentCells[attributes.indexPath.key()]
            cell?.let {
                this@LNCollectionView.removeView(it)
                currentCells.remove(attributes.indexPath.key())
                pool.addReusableCell(it)
            }
        }

        currentAttributesArr = newAttributesSet.toList()
        contentSize = collectionViewLayout.collectionViewContentSize
    }

    fun registerClassForCellWithReuseIdentifier(cellClass: Class<out LNCollectionViewCell>, identifier: String) {
        pool.registerClassForCellWithReuseIdentifier(cellClass, identifier)
    }

    fun dequeueReusableCellWithReuseIdentifier(identifier: String, indexPath: LNIndexPath): LNCollectionViewCell? {
        return pool.dequeueReusableCellWithIdentifier(identifier)
    }
}

